File: backend/app/routers/ai.py
import os
import uuid as uuid_mod

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings
from app.db.session import get_db
from app.deps import get_current_user
from app.models.user import User
from app.services.key_manager import get_user_key

router = APIRouter(prefix="/ai", tags=["ai"])


# ── Request schemas ──


class ChatMessage(BaseModel):
    role: str
    content: str


class ChatCompletionRequest(BaseModel):
    messages: list[ChatMessage]
    model: str | None = "deepseek-chat"
    temperature: float | None = 0.7
    maxTokens: int | None = 2000


class TTSRequest(BaseModel):
    text: str
    voice: str | None = "en-US-AriaNeural"
    model: str | None = "edge-tts"
    speed: float | None = 1.0


class STTRequest(BaseModel):
    audioUrl: str
    language: str | None = "en"
    model: str | None = "whisper-large-v3"


class ImageGenerationRequest(BaseModel):
    prompt: str
    negativePrompt: str | None = None
    model: str | None = "sdxl"
    width: int | None = 1024
    height: int | None = 1024


class MusicGenerationRequest(BaseModel):
    prompt: str
    duration: int | None = 30
    genre: str | None = None
    mood: str | None = None


class SFXRequest(BaseModel):
    prompt: str
    duration: float | None = None


class VideoGenerationRequest(BaseModel):
    prompt: str
    imageUrl: str | None = None
    duration: int | None = 4
    fps: int | None = 24
    width: int | None = 1280
    height: int | None = 720


class CaptionsRequest(BaseModel):
    audioUrl: str
    language: str | None = "en"


class VideoAnalyzeRequest(BaseModel):
    videoUrl: str


# ── Routes ──


@router.get("/capabilities")
async def get_capabilities():
    return {
        "textGeneration": {
            "chatCompletion": True,
            "streaming": False,
            "functionCalling": False,
            "vision": False,
        },
        "speech": {
            "textToSpeech": True,
            "speechToText": True,
            "voiceCloning": False,
            "audioGeneration": True,
        },
        "vision": {
            "imageGeneration": True,
            "imageToImage": False,
            "imageAnalysis": False,
            "backgroundRemoval": False,
        },
        "video": {
            "videoGeneration": False,
            "videoToVideo": False,
            "motionTracking": False,
            "sceneDetection": False,
        },
        "animation": {
            "lottieGeneration": False,
            "animationGeneration": False,
        },
    }


@router.get("/models")
async def get_models():
    return {
        "chat": ["deepseek-chat", "deepseek-flash"],
        "tts": ["edge-tts"],
        "stt": ["whisper-large-v3"],
        "image": ["sdxl"],
    }


@router.post("/chat/completion")
async def chat_completion(
    body: ChatCompletionRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    api_key = await get_user_key(db, user.id, "deepseek")
    if not api_key:
        raise HTTPException(status_code=400, detail="DeepSeek API key not configured. Please add it in Settings.")

    from openai import OpenAI

    client = OpenAI(api_key=api_key, base_url="https://api.deepseek.com")

    try:
        response = client.chat.completions.create(
            model=body.model or "deepseek-chat",
            messages=[{"role": m.role, "content": m.content} for m in body.messages],
            temperature=body.temperature or 0.7,
            max_tokens=body.maxTokens or 2000,
        )
        return {
            "content": response.choices[0].message.content,
            "usage": {
                "promptTokens": response.usage.prompt_tokens,
                "completionTokens": response.usage.completion_tokens,
                "totalTokens": response.usage.total_tokens,
            },
        }
    except Exception as e:
        if "invalid_api_key" in str(e).lower() or "authentication" in str(e).lower():
            raise HTTPException(status_code=400, detail="Invalid DeepSeek API key. Please check your settings.")
        raise HTTPException(status_code=500, detail=f"DeepSeek error: {str(e)}")


@router.post("/tts")
async def text_to_speech(
    body: TTSRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # Check for ElevenLabs preference
    if body.model == "elevenlabs":
        api_key = await get_user_key(db, user.id, "elevenlabs")
        if api_key:
            # Use ElevenLabs (premium, user-provided key)
            import httpx

            voice_id = body.voice or "21m00Tcm4TlvDq8ikWAM"
            async with httpx.AsyncClient() as http:
                resp = await http.post(
                    f"https://api.elevenlabs.io/v1/text-to-speech/{voice_id}",
                    headers={"xi-api-key": api_key, "Content-Type": "application/json"},
                    json={"text": body.text, "model_id": "eleven_monolingual_v1"},
                    timeout=60,
                )
                if resp.status_code != 200:
                    raise HTTPException(status_code=resp.status_code, detail="ElevenLabs TTS failed")

                filename = f"tts_{uuid_mod.uuid4()}.mp3"
                filepath = os.path.join(settings.upload_dir, filename)
                with open(filepath, "wb") as f:
                    f.write(resp.content)

                return {"audioUrl": f"/uploads/{filename}"}

    # Default: FREE Edge TTS
    import edge_tts

    voice = body.voice or "en-US-AriaNeural"
    communicate = edge_tts.Communicate(body.text, voice)

    filename = f"tts_{uuid_mod.uuid4()}.mp3"
    filepath = os.path.join(settings.upload_dir, filename)
    await communicate.save(filepath)

    return {"audioUrl": f"/uploads/{filename}"}


@router.post("/stt")
async def speech_to_text(
    body: STTRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # User key first, then server-side fallback
    api_key = await get_user_key(db, user.id, "groq") or settings.groq_api_key
    if not api_key:
        raise HTTPException(status_code=400, detail="Groq API key not configured. Please add it in Settings.")

    from groq import Groq

    client = Groq(api_key=api_key)

    try:
        # Download the audio if it's a URL
        import httpx

        audio_path = os.path.join(settings.upload_dir, f"stt_{uuid_mod.uuid4()}.mp3")
        async with httpx.AsyncClient() as http:
            resp = await http.get(body.audioUrl, timeout=60)
            with open(audio_path, "wb") as f:
                f.write(resp.content)

        with open(audio_path, "rb") as audio_file:
            transcription = client.audio.transcriptions.create(
                model="whisper-large-v3",
                file=audio_file,
                response_format="verbose_json",
                language=body.language or "en",
            )

        os.remove(audio_path)

        return {
            "text": transcription.text,
            "segments": [
                {"start": seg.start, "end": seg.end, "text": seg.text}
                for seg in (transcription.segments or [])
            ],
        }
    except Exception as e:
        if "authentication" in str(e).lower():
            raise HTTPException(status_code=400, detail="Invalid Groq API key.")
        raise HTTPException(status_code=500, detail=f"Transcription error: {str(e)}")


@router.post("/image/generate")
async def generate_image(
    body: ImageGenerationRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # User key first, then server-side fallback
    api_key = await get_user_key(db, user.id, "replicate") or settings.replicate_api_token
    if not api_key:
        raise HTTPException(status_code=400, detail="Replicate API token not configured. Please add it in Settings.")

    import replicate

    client = replicate.Client(api_token=api_key)

    try:
        output = client.run(
            "stability-ai/sdxl:39ed52f2a78e934b3ba6e2a89f5b1c712de7dfea535525255b1aa35c5565e08b",
            input={
                "prompt": body.prompt,
                "negative_prompt": body.negativePrompt or "",
                "width": body.width or 1024,
                "height": body.height or 1024,
                "num_outputs": 1,
            },
        )
        image_url = output[0] if isinstance(output, list) else str(output)
        return {"imageUrl": image_url}
    except Exception as e:
        if "authentication" in str(e).lower():
            raise HTTPException(status_code=400, detail="Invalid Replicate API token.")
        raise HTTPException(status_code=500, detail=f"Image generation error: {str(e)}")


@router.post("/music/generate")
async def generate_music(
    body: MusicGenerationRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # Placeholder - connect to your preferred music generation service
    raise HTTPException(status_code=501, detail="Music generation not yet configured. Connect a service in Settings.")


@router.post("/sfx/generate")
async def generate_sfx(
    body: SFXRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # Placeholder - connect to ElevenLabs SFX or similar
    api_key = await get_user_key(db, user.id, "elevenlabs")
    if not api_key:
        raise HTTPException(status_code=400, detail="ElevenLabs API key not configured for SFX generation.")

    raise HTTPException(status_code=501, detail="SFX generation endpoint - implement with ElevenLabs SFX API")


@router.post("/video/generate")
async def generate_video(
    body: VideoGenerationRequest,
    user: User = Depends(get_current_user),
):
    # Placeholder for video generation (Runway, Pika, etc.)
    raise HTTPException(status_code=501, detail="Video generation not yet configured.")


@router.get("/video/status/{job_id}")
async def check_video_status(job_id: str, user: User = Depends(get_current_user)):
    # Placeholder for async video job status
    return {"status": "pending", "videoUrl": None}


@router.get("/lottie/search")
async def search_lottie(q: str = ""):
    # Placeholder for Lottie search
    return {"results": []}


@router.post("/captions/generate")
async def generate_captions(
    body: CaptionsRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # User key first, then server-side fallback
    api_key = await get_user_key(db, user.id, "groq") or settings.groq_api_key
    if not api_key:
        raise HTTPException(status_code=400, detail="Groq API key not configured.")

    from groq import Groq

    client = Groq(api_key=api_key)

    try:
        import httpx

        audio_path = os.path.join(settings.upload_dir, f"caption_{uuid_mod.uuid4()}.mp3")
        async with httpx.AsyncClient() as http:
            resp = await http.get(body.audioUrl, timeout=60)
            with open(audio_path, "wb") as f:
                f.write(resp.content)

        with open(audio_path, "rb") as audio_file:
            transcription = client.audio.transcriptions.create(
                model="whisper-large-v3",
                file=audio_file,
                response_format="verbose_json",
                language=body.language or "en",
            )

        os.remove(audio_path)

        return {
            "segments": [
                {"start": seg.start, "end": seg.end, "text": seg.text}
                for seg in (transcription.segments or [])
            ]
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Caption generation error: {str(e)}")


@router.post("/video/analyze")
async def analyze_video(body: VideoAnalyzeRequest, user: User = Depends(get_current_user)):
    # Placeholder for video analysis
    raise HTTPException(status_code=501, detail="Video analysis not yet configured.")


@router.get("/usage")
async def get_usage(
    start: str | None = None,
    end: str | None = None,
    user: User = Depends(get_current_user),
):
    # Placeholder for usage tracking
    return []


@router.get("/tts/voices")
async def get_tts_voices():
    """Return available Edge TTS voices."""
    import edge_tts

    voices = await edge_tts.list_voices()
    return {
        "voices": [
            {
                "id": v["ShortName"],
                "name": v["FriendlyName"],
                "locale": v["Locale"],
                "gender": v["Gender"],
            }
            for v in voices
            if v["Locale"].startswith("en-")
        ]
    }

File: src/lib/api/ai.ts
// ── AI Services (local) ──────────────────────────────────────────────────
//
// Guide Studio runs its AI locally. This module keeps the request/response
// shapes callers were written against, but every call now goes over the
// Electron IPC bridge to the local AI service (Ollama or the user's own
// provider key, configured in Settings) instead of the Guide Studio backend.
//
// Local providers behind each method:
//   chatCompletion  — ai-analyze / ai-analyze-image
//   generateSpeech  — ai-minimax-tts, falling back to ai-tts-synthesize
//   listVoices      — ai-minimax-voices
//   generateImage   — ai-minimax-image
//   generateMusic   — ai-generate-music
//   transcribe      — not available here; captions use the local Whisper
//                     engine in `src/lib/captioning/transcribe.ts`

// ── Supported AI Capabilities ────────────────────────────────────────────

export interface AICapabilities {
	// Text Generation (GPT-4, Claude, Gemini, etc.)
	textGeneration: {
		chatCompletion: boolean;
		streaming: boolean;
		functionCalling: boolean;
		vision: boolean;
	};

	// Speech & Audio
	speech: {
		textToSpeech: boolean; // TTS (ElevenLabs, OpenAI TTS, etc.)
		speechToText: boolean; // STT/Whisper
		voiceCloning: boolean;
		audioGeneration: boolean; // Music, SFX (ElevenLabs, Suno, etc.)
	};

	// Vision & Image
	vision: {
		imageGeneration: boolean; // DALL-E, Midjourney, Stable Diffusion
		imageToImage: boolean; // Style transfer, upscaling
		imageAnalysis: boolean; // Vision models
		backgroundRemoval: boolean;
	};

	// Video
	video: {
		videoGeneration: boolean; // Runway, Pika, etc.
		videoToVideo: boolean; // Style transfer
		motionTracking: boolean;
		sceneDetection: boolean;
	};

	// Animation
	animation: {
		lottieGeneration: boolean; // Animated icons/graphics
		animationGeneration: boolean; // Motion graphics
	};
}

// ── AI Service Interfaces ────────────────────────────────────────────────

// Chat/Text Generation
/** OpenAI-style multimodal content part. Messages containing image parts are
 *  routed to the backend's vision model (Gemini) instead of DeepSeek. */
export interface ChatContentPart {
	type: "text" | "image_url";
	text?: string;
	image_url?: { url: string };
}

export interface ChatMessage {
	role: "system" | "user" | "assistant";
	content: string | ChatContentPart[];
}

export interface ChatCompletionRequest {
	messages: ChatMessage[];
	model?: string;
	temperature?: number;
	max_tokens?: number;
	stream?: boolean;
}

// Text-to-Speech
export interface TTSRequest {
	text: string;
	voice?: string; // Voice ID
	model?: string; // "edge-tts" or "elevenlabs"
	speed?: number;
	/**
	 * The language the narration is in, e.g. "de" or "pt-BR". Only the local
	 * Piper engine uses it, to pick a voice that speaks it; leave it out and the
	 * language is read from the text.
	 */
	language?: string;
}

// Image Generation
export interface ImageGenerationRequest {
	prompt: string;
	negative_prompt?: string;
	model?: string; // "sdxl" or "flux"
	width?: number;
	height?: number;
	num_outputs?: number;
}

// Music Generation
export interface MusicGenerationRequest {
	prompt: string;
	duration?: number; // seconds (5-30)
	model_version?: string;
}

// Video Generation
export interface VideoGenerationRequest {
	prompt: string;
	imageUrl?: string; // Optional starting image
	duration?: number; // seconds
	fps?: number;
	width?: number;
	height?: number;
}

// ── AI Service Class ──────────────────────────────────────────────────────

/** Flatten an OpenAI-style message list into a prompt plus a system context. */
function splitMessages(messages: ChatMessage[]) {
	const system: string[] = [];
	const turns: string[] = [];
	let imageBase64: string | undefined;

	for (const message of messages) {
		const parts = typeof message.content === "string" ? [message.content] : message.content;
		const text: string[] = [];
		for (const part of parts) {
			if (typeof part === "string") {
				text.push(part);
				continue;
			}
			if (part.type === "text" && part.text) text.push(part.text);
			if (part.type === "image_url" && part.image_url?.url && !imageBase64) {
				// The local vision call takes bare base64, not a data URL.
				imageBase64 = part.image_url.url.replace(/^data:[^;]+;base64,/, "");
			}
		}
		const joined = text.join("\n").trim();
		if (!joined) continue;
		if (message.role === "system") system.push(joined);
		else turns.push(message.role === "assistant" ? `Assistant: ${joined}` : joined);
	}

	return {
		prompt: turns.join("\n\n"),
		context: system.length > 0 ? system.join("\n\n") : undefined,
		imageBase64,
	};
}

export class AIService {
	// Chat completion through the local provider. Messages carrying an image
	// part are routed to the local vision call instead.
	async chatCompletion(
		request: ChatCompletionRequest,
		_opts?: { timeoutMs?: number },
	): Promise<
		{ success: true; data: { content: string; usage: unknown } } | { success: false; error: string }
	> {
		const { prompt, context, imageBase64 } = splitMessages(request.messages);
		if (!window.electronAPI?.aiAnalyze) {
			return { success: false, error: "No local AI provider is available" };
		}

		try {
			const result = imageBase64
				? await window.electronAPI.aiAnalyzeImage(prompt, imageBase64, context)
				: await window.electronAPI.aiAnalyze(prompt, context);
			if (result.success && result.text) {
				return { success: true, data: { content: result.text, usage: null } };
			}
			return { success: false, error: result.error || "The local AI provider returned no text" };
		} catch (error) {
			return {
				success: false,
				error: error instanceof Error ? error.message : "Local AI request failed",
			};
		}
	}

	// Text-to-speech through the local TTS providers. Returns a local file
	// path, which both the preview <audio> elements and FFmpeg narration
	// muxing at export can use.
	async generateSpeech(
		request: TTSRequest,
	): Promise<{ success: true; data: { audioUrl: string } } | { success: false; error: string }> {
		try {
			if (window.electronAPI?.aiMinimaxTts) {
				const minimax = await window.electronAPI.aiMinimaxTts(request.text, {
					voiceId: request.voice,
					speed: request.speed,
				});
				if (minimax.success && minimax.audioPath) {
					return { success: true, data: { audioUrl: minimax.audioPath } };
				}
			}

			if (window.electronAPI?.aiTtsSynthesize) {
				const fallback = await window.electronAPI.aiTtsSynthesize(request.text, request.voice, {
					language: request.language,
				});
				if (fallback.success && fallback.audioPath) {
					return { success: true, data: { audioUrl: fallback.audioPath } };
				}
				return { success: false, error: fallback.error || "Local TTS failed" };
			}

			return { success: false, error: "No local text-to-speech provider is available" };
		} catch (error) {
			return {
				success: false,
				error: error instanceof Error ? error.message : "TTS request failed",
			};
		}
	}

	// List the local TTS voices.
	async listVoices(_locale?: string): Promise<
		| {
				success: true;
				data: { voices: Array<{ id: string; name: string; locale: string; gender: string }> };
		  }
		| { success: false; error: string }
	> {
		if (!window.electronAPI?.aiMinimaxVoices) {
			return { success: false, error: "No local text-to-speech provider is available" };
		}
		const result = await window.electronAPI.aiMinimaxVoices();
		return {
			success: true,
			data: {
				voices: result.voices.map((voice) => ({
					id: voice.id,
					name: voice.name,
					locale: "",
					gender: voice.gender,
				})),
			},
		};
	}

	// Speech-to-text. There is no local STT behind this call — auto-captions
	// run the in-app Whisper engine directly (`transcribeMono16kToSegments`).
	async transcribe(
		_audioFile: File | Blob,
		_language?: string,
	): Promise<
		| {
				success: true;
				data: {
					segments: Array<{ start: number; end: number; text: string }>;
					fullText: string;
					language: string;
				};
		  }
		| { success: false; error: string }
	> {
		return { success: false, error: "Transcription runs on the local caption engine" };
	}

	// Image generation through the local image provider.
	async generateImage(
		request: ImageGenerationRequest,
	): Promise<
		{ success: true; data: { images: Array<{ url: string }> } } | { success: false; error: string }
	> {
		if (!window.electronAPI?.aiMinimaxImage) {
			return { success: false, error: "No local image provider is available" };
		}
		const result = await window.electronAPI.aiMinimaxImage(request.prompt, {
			count: request.num_outputs,
		});
		if (result.success && result.imagePaths?.length) {
			return { success: true, data: { images: result.imagePaths.map((url) => ({ url })) } };
		}
		return { success: false, error: result.error || "Local image generation failed" };
	}

	// Music generation through the local music provider.
	async generateMusic(
		request: MusicGenerationRequest,
		_timeoutMs?: number,
	): Promise<
		| { success: true; data: { audioUrl: string; duration: number } }
		| { success: false; error: string }
	> {
		if (!window.electronAPI?.aiGenerateMusic) {
			return { success: false, error: "No local music provider is available" };
		}
		const result = await window.electronAPI.aiGenerateMusic(
			"custom",
			request.prompt,
			request.duration,
		);
		if (result.success && result.audioPath) {
			return {
				success: true,
				data: { audioUrl: result.audioPath, duration: result.durationSec ?? request.duration ?? 0 },
			};
		}
		return { success: false, error: result.error || "Local music generation failed" };
	}

	// SFX generation through the local SFX provider.
	async generateSFX(
		prompt: string,
		duration?: number,
	): Promise<{ success: true; data: { audioUrl: string } } | { success: false; error: string }> {
		if (!window.electronAPI?.aiElevenlabsSfx) {
			return { success: false, error: "No local sound-effect provider is available" };
		}
		const result = await window.electronAPI.aiElevenlabsSfx(prompt, { durationSec: duration });
		if (result.success && result.filePath) {
			return { success: true, data: { audioUrl: result.filePath } };
		}
		return { success: false, error: result.error || "Local sound-effect generation failed" };
	}

	// Video Generation (no local provider)
	async generateVideo(
		_request: VideoGenerationRequest,
	): Promise<
		| { success: true; data: { videoUrl: string; jobId?: string } }
		| { success: false; error: string }
	> {
		return { success: false, error: "Video generation is not available" };
	}

	// Check video generation status
	async checkVideoStatus(_jobId: string): Promise<
		| {
				success: true;
				data: { status: "pending" | "processing" | "completed" | "failed"; videoUrl?: string };
		  }
		| { success: false; error: string }
	> {
		return { success: false, error: "Video generation is not available" };
	}

	// Lottie Animation Search (served by the local Lottie search handlers)
	async searchLottie(
		_query: string,
	): Promise<
		| { success: true; data: { results: Array<{ id: string; url: string; preview: string }> } }
		| { success: false; error: string }
	> {
		return { success: false, error: "Lottie search is not available here" };
	}

	// Auto-caption generation (uses the local caption engine under the hood)
	async generateCaptions(
		_audioUrl: string,
		_language?: string,
	): Promise<
		| { success: true; data: { segments: Array<{ start: number; end: number; text: string }> } }
		| { success: false; error: string }
	> {
		return { success: false, error: "Captions run on the local caption engine" };
	}

	// AI Video Analysis (no local provider)
	async analyzeVideo(
		_videoUrl: string,
	): Promise<
		| { success: true; data: { scenes: unknown[]; keyframes: unknown[]; summary: string } }
		| { success: false; error: string }
	> {
		return { success: false, error: "Video analysis is not available" };
	}
}

export const aiService = new AIService();

// ── Usage Token Tracking ──────────────────────────────────────────────────

export interface TokenUsage {
	service: string;
	tokens: number;
	cost: number;
	timestamp: string;
}

/** Local usage tracking is reported by the main process, not an account. */
export async function getTokenUsage(
	_startDate?: string,
	_endDate?: string,
): Promise<{ success: true; data: TokenUsage[] } | { success: false; error: string }> {
	return { success: true, data: [] };
}
